daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala

```diff
@@ -202,6 +212,29 @@ final case class DefTemplate[+Ty](
   }
 
   def getKey: j.Optional[_ <: Ty] = toOptional(key)
+
+  private[iface] def extendWithInterface[OTy >: Ty](
+      ifaceName: Ref.TypeConName,
+      ifc: DefInterface[OTy],
+  ): DefTemplate[OTy] = {
+    import TemplateChoices.{Resolved, Unresolved}
+    copy(
+      implementedInterfaces = implementedInterfaces :+ ifaceName,
+      tChoices = tChoices match {
+        case unr @ Unresolved(_, sources) =>
+          unr.copy(unresolvedChoiceSources = sources incl ifaceName)
+        // If unresolved, we need only add ifc as a future interface to resolve;
+        // otherwise, we must self-resolve and add to preexisting resolutions
+        case r @ Resolved(rc) =>
+          type K[C] = Semigroup[Resolved.Choices[C]]
+          r.copy(resolvedChoices =
+            FirstVal
+              .unsubst[K, TemplateChoice[OTy]](Semigroup.apply)
+              .append(rc, ifc choicesAsResolved ifaceName)
```

Comment:
I'm assuming the choices resolved with if and ifaceName are appended to rc, but I don't understand why the Firstval, semigroup stuff is needed.

Reply:
The semigroup is the same one used for aggregating interface choices [when doing choice resolution](https://github.com/digital-asset/daml/blob/8d21d19133a1ca507760fb1b875ece3d6ad2d6db/daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala#L287-L301). So we're just doing one step of that aggregation here.